metrics/src/poh_timing_point.rs

```diff
@@ -0,0 +1,207 @@
+        }
+    }
+}
+
+/// SlotPohTimingInfo. This struct is sent to channel and received by
+/// poh_timing_report service.
+#[derive(Clone, Debug)]
+pub struct SlotPohTimingInfo {
+    /// current slot
+    pub slot: Slot,
+    /// root slot
+    pub root_slot: Option<Slot>,
+    /// timing event
+    pub timing_point: PohTimingPoint,
+}
+
+impl fmt::Display for SlotPohTimingInfo {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "PohTimingPoint: {}, slot={}, root_slot={}",
+            self.timing_point,
+            self.slot,
+            self.root_slot.unwrap_or(0),
+        )
+    }
+}
+
+/// overload parameters for SlotTimingInfo constructor
+pub enum SlotTimingInfoParams {
+    /// (slot, timestamp)
+    WithoutRoot(Slot, u64),
+
+    /// (slot, root_slot, timestamp)
+    WithRoot(Slot, Slot, u64),
+}
+
+/// overload traits
+pub trait IntoSlotTimingInfoParams {
+    fn into(self) -> SlotTimingInfoParams;
+}
+
+/// WithoutRoot
+impl IntoSlotTimingInfoParams for (Slot, u64) {
+    fn into(self) -> SlotTimingInfoParams {
+        SlotTimingInfoParams::WithoutRoot(self.0, self.1)
+    }
+}
+
+/// WithRoot
+impl IntoSlotTimingInfoParams for (Slot, Slot, u64) {
+    fn into(self) -> SlotTimingInfoParams {
+        SlotTimingInfoParams::WithRoot(self.0, self.1, self.2)
+    }
+}
+
+impl SlotPohTimingInfo {
+    /// create slot poh timing point w/o root info.
+    fn new_slot_poh_time_point<A, F>(args: A, f: F) -> SlotPohTimingInfo
+    where
+        A: IntoSlotTimingInfoParams,
+        F: FnOnce(u64) -> PohTimingPoint,
+    {
+        let param = args.into();
+        match param {
+            SlotTimingInfoParams::WithoutRoot(slot, timestamp) => SlotPohTimingInfo {
+                slot,
+                root_slot: None,
+                timing_point: f(timestamp),
+            },
+            SlotTimingInfoParams::WithRoot(slot, root_slot, timestamp) => SlotPohTimingInfo {
+                slot,
+                root_slot: Some(root_slot),
+                timing_point: f(timestamp),
+            },
+        }
+    }
+
+    /// create slot start poh timing point w/o root info.
+    pub fn new_slot_start_poh_time_point<A>(args: A) -> SlotPohTimingInfo
+    where
+        A: IntoSlotTimingInfoParams,
+    {
+        SlotPohTimingInfo::new_slot_poh_time_point(args, PohTimingPoint::PohSlotStart)
+    }
+
+    /// create slot end poh timing point w/o root info.
+    pub fn new_slot_end_poh_time_point<A>(args: A) -> SlotPohTimingInfo
+    where
+        A: IntoSlotTimingInfoParams,
+    {
+        SlotPohTimingInfo::new_slot_poh_time_point(args, PohTimingPoint::PohSlotEnd)
+    }
+
+    /// create slot full poh timing point w/o root info.
+    pub fn new_slot_full_poh_time_point<A>(args: A) -> SlotPohTimingInfo
+    where
+        A: IntoSlotTimingInfoParams,
+    {
+        SlotPohTimingInfo::new_slot_poh_time_point(args, PohTimingPoint::FullSlotReceived)
+    }
+}
```

Comment:
I think these constructor arguments are actually overcomplicated, i.e. the generic `IntoSlotTimingInfoParams` is unnecessary. I think having the user explictly pass in the `slot: Slot` and `root_slot: Option<Slot>` into the set of functions:

```
new_slot_full_poh_time_point(slot: Slot, root_slot: Option<Slot>)
new_slot_end_poh_time_point(slot: Slot, root_slot: Option<Slot>)
new_slot_start_poh_time_point(slot: Slot, root_slot: Option<Slot>)
```

This makes it easier to read what these arguments actually are

Reply:
OK. I will modify.

Reply:
Pushed.